sdks/python/apache_beam/io/restriction_trackers.py

```diff
@@ -0,0 +1,131 @@
+    while current_split_start < self.stop:
+      current_split_stop = min(current_split_start + max_split_size, self.stop)
+      remaining = self.stop - current_split_stop
+
+      # Avoiding a small split at the end.
+      if (remaining < desired_num_offsets_per_split / 4 or
+          remaining < min_num_offsets_per_split):
+        current_split_stop = self.stop
+
+      yield OffsetRange(current_split_start, current_split_stop)
+      current_split_start = current_split_stop
+
+  def new_tracker(self):
+    return OffsetRangeTracker(self.start, self.stop)
+
+
+class OffsetRestrictionTracker(RestrictionTracker):
+  """An `iobase.RestrictionTracker` implementations for an offset range.
+
+  Offset range is represented as a pair of integers
+  [start_position, stop_position}.
+  """
+
+  def __init__(self, start_position, stop_position):
+    self._range = OffsetRange(start_position, stop_position)
+    self._current_position = None
+    self._last_claim_attempt = None
+    self._checkpointed = False
+    self._lock = threading.Lock()
+
+  def check_done(self):
+    with self._lock:
+      if self._last_claim_attempt < self._range.stop - 1:
+        raise ValueError(
+            'OffsetRestrictionTracker is not done since work in range [%s, %s) '
+            'has not been claimed.',
+            self._last_claim_attempt if self._last_claim_attempt is not None
+            else self._range.start,
+            self._range.stop)
+
+  def current_restriction(self):
+    with self._lock:
+      return (self._range.start, self._range.stop)
+
+  def start_position(self):
+    with self._lock:
+      return self._range.start
+
+  def stop_position(self):
+    with self._lock:
+      return self._range.stop
+
+  def try_claim(self, position):
+    with self._lock:
+      if self._last_claim_attempt and position <= self._last_claim_attempt:
+        raise ValueError(
+            'Positions claimed should strictly increase. Trying to claim '
+            'position %d while last claim attempt was %d.',
+            position, self._last_claim_attempt)
+
+      self._last_claim_attempt = position
+      if position < self._range.start:
+        raise ValueError(
+            'Position to be claimed cannot be smaller than the start position '
+            'of the range. Tried to claim position %r for the range [%r, %r)',
+            position, self._range.start, self._range.stop)
+
+      if position >= self._range.start and position < self._range.stop:
+        self._current_position = position
+        return True
+
+      return False
+
+  def checkpoint(self):
+    with self._lock:
+      # If self._current_position is 'None' no records have been claimed so
+      # residual should start from self._range.start.
+      if self._current_position is None:
+        residual_range = (self._range.start, self._range.stop)
+        end_position = self._range.start
+      else:
+        residual_range = (self._current_position + 1, self._range.stop)
```

Comment:
<!--new_thread; commit:3d6511ca1799a602673dbf658e25d6877931a944; resolved:0-->
Can you factor `residual_range` out of this if?  Just as:

```python
residual_range = (end_position, self._range.stop)
```